### PR TITLE
Feat: Update rarity functions. Add tests. Add on chain function

### LINKED ISSUE #21: As a User I want to calculate rarity for a collection
The whole functionality should already be implemented in the `hedera-nft-utilities` repository (https://github.com/hashgraph/hedera-nft-utilities?tab=readme-ov-file#rarity-score-calculation)

**Acceptance criteria:**
- The user should get a rarity ranking for each processed NFT
- It should work for on-chain and off-chain collections
  - This can be done by creating two separate functions:
    - Download collection metadata to memory
    - Calculate off-chain rarity
- Unit tests are implemented

### PR DESCRIPTION
**Pull request by StanislawMazowieckiAriane**
Fixes #21

### CODE REVIEW

**Review by BartoszSolkaArianelabs — APPROVED** [review]

File: src/rarity/index.ts [hunk 1]
```diff
@@ -58,23 +56,16 @@ const calculateRarity = (dir: string): RarityResult[] => {
...
-        attributeConfigObject.values.find(
-          (valueObject) => valueObject.value === NFTAttribute.value
-        );
-      const mostCommonTrait = attributeConfigObject.values.reduce(
-        (prev, current) => (prev.count > current.count ? prev : current)
+      const attributeConfigObject: AttributeConfig | undefined = attributesMap.find(
+        (attribute) => attribute.trait_type === NFTAttribute.trait_type
+      );
+
+      if (!attributeConfigObject) throw new Error(`Attribute ${NFTAttribute.trait_type} not found in attributes map`);
```
Comment (michalrozekariane) [inline src/rarity/index.ts:64]: pls add this error to dictionary

File: src/rarity/index.ts [hunk 2]
```diff
@@ -111,24 +102,17 @@ const getAttributeMap = (files: NFTFile[]): AttributeConfig[] => {
   const attributesMap: AttributeConfig[] = [];
   files.forEach((file) => {
     if (!file.filedata.attributes)
-      throw new Error(
-        `Attributes not found in file ${file.filename}. Please ensure that your metadata file is valid.`
-      );
+      throw new Error(`Attributes not found in file ${file.filename}. Please ensure that your metadata file is valid.`);
```
Comment (michalrozekariane) [inline src/rarity/index.ts:105]: pls add this error to dictionary

File: src/rarity/index.ts [hunk 3]
```diff
@@ -151,28 +135,21 @@ const getAttributeMap = (files: NFTFile[]): AttributeConfig[] => {
...
  * */
-const getAttributeMapData = (metadataArray: { attributes: Attribute[] }[]): AttributeConfig[] => {
+const getAttributeMapData = (metadataArray: MetadataObject[]): AttributeConfig[] => {
   const attributesMap: AttributeConfig[] = [];
   metadataArray.forEach((metadata) => {
     if (!metadata.attributes)
-      throw new Error(
-        `Attributes not found in object ${JSON.stringify(metadata)}. Please ensure that your metadata file is valid.`
-      );
+      throw new Error(`Attributes not found in object ${JSON.stringify(metadata)}. Please ensure that your metadata file is valid.`);
```
Comment (michalrozekariane) [inline src/rarity/index.ts:142]: pls use error from dictionary

File: src/rarity/index.ts [hunk 4]
```diff
@@ -196,31 +173,24 @@ const getAttributeMapData = (metadataArray: { attributes: Attribute[] }[]): Attr
...
-        throw new Error(
-          `Attribute ${NFTAttribute.trait_type} not found in attributes map`
-        );
-      const NFTsWithTrait: ValueObject | undefined =
-        attributeConfigObject.values.find(
-          (valueObject) => valueObject.value === NFTAttribute.value
-        );
-      const mostCommonTrait = attributeConfigObject.values.reduce(
-        (prev, current) => (prev.count > current.count ? prev : current)
+      if (!attributeConfigObject) throw new Error(`Attribute ${NFTAttribute.trait_type} not found in attributes map`);
```
Comment (michalrozekariane) [inline src/rarity/index.ts:189]: pls use error from dictionary

**Review by michalrozekariane — CHANGES_REQUESTED** [review]

**Review by se7enarianelabs — APPROVED** [review]

File: src/rarity/index.ts (hunk 3 above)
Comment (StanislawMazowieckiAriane) [reply src/rarity/index.ts]: Done

File: src/rarity/index.ts (hunk 4 above)
Comment (StanislawMazowieckiAriane) [reply src/rarity/index.ts]: Done

File: src/rarity/index.ts (hunk 1 above)
Comment (StanislawMazowieckiAriane) [reply src/rarity/index.ts]: Done

**Review by michalrozekariane — APPROVED** [review]

**Pull request merged**